zkevm-circuits/src/evm_circuit/execution/codecopy.rs

```diff
@@ -0,0 +1,476 @@
+                    next_bytes_left.expr(),
+                    dst_memory_addr.length(),
+                );
+                cb.require_equal(
+                    "next_src_addr_end == code_size",
+                    next_src_addr_end.expr(),
+                    code_size.expr(),
+                );
+                cb.require_equal(
+                    "next_code_hash == code_hash",
+                    next_code_hash.expr(),
+                    code_hash.expr(),
+                );
+            },
+        );
+
+        // Expected state transition.
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(cb.rw_counter_offset()),
+            program_counter: Transition::Delta(1.expr()),
+            stack_pointer: Transition::Delta(3.expr()),
+            memory_word_size: Transition::To(memory_expansion.next_memory_word_size()),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            Some(memory_copier_gas.gas_cost()),
+        );
+
+        Self {
+            same_context,
+            code_offset,
+            code_size,
+            account,
+            dst_memory_addr,
+            memory_expansion,
+            memory_copier_gas,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction<F>,
+        call: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        // 1. `dest_offset` is the bytes offset in the memory where we start to
+        // write.
+        // 2. `code_offset` is the bytes offset in the current
+        // context's code where we start to read.
+        // 3. `size` is the number of
+        // bytes to be read and written (0s to be copied for out of bounds).
+        let [dest_offset, code_offset, size] =
+            [0, 1, 2].map(|i| block.rws[step.rw_indices[i]].stack_value());
+
+        // assign the code offset memory address.
+        self.code_offset.assign(
+            region,
+            offset,
+            Some(
+                code_offset.to_le_bytes()[..N_BYTES_MEMORY_ADDRESS]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+
+        let account = if call.is_root {
+            call.callee_address
+        } else {
+            unimplemented!("CODECOPY does not support internal calls yet");
+        };
+        self.account.assign(
+            region,
+            offset,
+            Some(
+                account
+                    .to_scalar()
+                    .expect("could not convert address to scalar field"),
+            ),
+        )?;
+
+        let code = block
+            .bytecodes
+            .iter()
+            .find(|b| {
```

Comment:
This seems fine for now, but we should consider changing the `Block` witness from
```
pub struct Block<F> {
    ...
    /// Bytecode used in the block
    pub bytecodes: Vec<Bytecode>,
    ...
}
``` 
to 
```
pub struct Block<F> {
    ...
    /// Bytecode used in the block
    pub bytecodes: HashMap<Word, Vec<u8>>,
    ...
}
```
so we don't end up having to find the same bytecode over and over again.

Reply:
Yea, I think we can address this in another PR, while also refactoring other parts of the codebase that make use of `block.bytecodes`